test/utils.spec.js

```diff
@@ -117,5 +117,86 @@ describe("Utils", () => {
         });
       });
     });
+    describe("Nested fields extraction", () => {
+      it("extracts nested fields using dot notation", async () => {
+        const typesenseDocumentFromSnapshot = (await import("../functions/src/utils.js")).typesenseDocumentFromSnapshot;
+        const documentSnapshot = test.firestore.makeDocumentSnapshot(
+          {
+            user: {
+              name: "John Doe",
+              address: {
+                city: "New York",
+                country: "USA",
+              },
+            },
+            tags: ["tag1", "tag2"],
+          },
+          "id",
+        );
+        const result = await typesenseDocumentFromSnapshot(documentSnapshot, ["user.name", "user.address.city", "tags"]);
+        expect(result).toEqual({
+          id: "id",
+          "user.name": "John Doe",
+          "user.address.city": "New York",
```

Comment:
@tharropoulos We seem to be flattening the key names into a dot notation. Can we retain the same nested structure as in the original field value, in the typesense document?

Reply:
Should we retain the last key of the nesting or use  snake case / camel case for them?

Reply:
We want to keep the same key names and nesting structure. So for eg, say the user specifies `user.address.city` as the only field to sync over, then the Typesense document would look like this:

```js
{
    id: "id",
    user: {
        address: {
            city: "New York",
        },
    },
}
```

Notice how `user.name` is not present in the nested structure. And the city nested field is still nested under address and user.

Reply:
> We want to keep the same key names and nesting structure. So for eg, say the user specifies `user.address.city` as the only field to sync over, then the Typesense document would look like this:
> 
> ```js
> {
>     id: "id",
>     user: {
>         address: {
>             city: "New York",
>         },
>     },
> }
> ```
> 
> Notice how `user.name` is not present in the nested structure. And the city nested field is still nested under address and user.

```javascript
        const result = await typesenseDocumentFromSnapshot(documentSnapshot, ["comments[0].author", "comments[1].text"]);
        expect(result).toEqual({
          id: "id",
          "comments[0].author": "Alice",
          "comments[1].text": "Thanks for sharing.",
        });
```

What about here? Do we drop indexed support altogether or concat them into 

```javascript
comments: {
    author:  "Alice",
    text: "Thanks for sharing"
}
```

I personally think it shouldn't use indexed values,  so I omitted the test case and refactored it to more closely match real world scenarios on f3fd3e8 .

Reply:
I would say we should retain the same data types in the transformed data as the source. 

So in the example you used, the transformed document should look like this:

```json
{
  "id": "id",
  "comments": {
    "author": ["Alice"],
    "text": ["Great post!"]
  }
}
```